catalog/harvest/catalog-harvest-metacardadapter/src/main/java/org/codice/ddf/catalog/harvest/adapter/HarvestedResourceTransformer.java

```diff
@@ -0,0 +1,189 @@
+    this.mimeTypeMapper = mimeTypeMapper;
+  }
+
+  /**
+   * Creates a metacard.
+   *
+   * @param harvestedResource the {@link HarvestedResource} the metacard will be created from
+   * @return the metacard, or null if there was an error
+   */
+  public Metacard transformHarvestedResource(HarvestedResource harvestedResource) {
+    return transform(harvestedResource, null);
+  }
+
+  /**
+   * Creates a metacard with an update id.
+   *
+   * @param resource the {@link HarvestedResource} the metacard will be created from
+   * @param metacardId id of the existing metacard to update
+   * @return the metacard, or null if there was an error
+   */
+  public Metacard transformHarvestedResource(HarvestedResource resource, String metacardId) {
+    return transform(resource, metacardId);
+  }
+
+  private Metacard transform(HarvestedResource harvestedResource, String metacardId) {
+    MimeType resourceMimeType = harvestedResource.getMimeType();
+
+    Metacard metacard = null;
+
+    try (TemporaryFileBackedOutputStream tfbos = new TemporaryFileBackedOutputStream()) {
+      IOUtils.copy(harvestedResource.getInputStream(), tfbos);
+      if (resourceMimeType == null) {
+        resourceMimeType =
+            guessMimeTypeFor(
+                tfbos.asByteSource().openStream(),
+                FilenameUtils.getExtension(harvestedResource.getName()));
+      }
+
+      metacard =
+          doTransform(
+              resourceMimeType, tfbos.asByteSource().openStream(), harvestedResource, metacardId);
+
+      if (metacard != null) {
+        enrichMetacard(metacard, harvestedResource);
+        return metacard;
+      }
+    } catch (IOException e) {
+      LOGGER.debug("Failed to open TFBOS for harvested resource's input stream.", e);
+    }
+    return metacard;
+  }
+
+  private void enrichMetacard(Metacard metacard, HarvestedResource harvestedResource) {
+    writeMetacardAttribute(metacard, Core.TITLE, harvestedResource.getName());
+    writeMetacardAttribute(
+        metacard, Core.RESOURCE_SIZE, Long.toString(harvestedResource.getSize()));
+    writeMetacardAttribute(metacard, Core.RESOURCE_URI, harvestedResource.getUri().toASCIIString());
+  }
+
+  private Metacard doTransform(
+      MimeType mimeType, InputStream is, Resource resource, String metacardId) {
+    List<InputTransformer> transformerCandidates =
+        mimeTypeToTransformerMapper.findMatches(InputTransformer.class, mimeType);
+
+    for (InputTransformer inputTransformer : transformerCandidates) {
+      Optional<Metacard> metacardOptional =
+          doTransform(inputTransformer, metacardId, is, resource.getName());
+
+      if (metacardOptional.isPresent()) {
+        return metacardOptional.get();
+      }
+    }
+    return null;
+  }
+
+  private Optional<Metacard> doTransform(
+      InputTransformer inputTransformer, String metacardId, InputStream is, String resourceName) {
+    try {
+      if (StringUtils.isNotEmpty(metacardId)) {
```

Comment:
❓ Kind of bugs me that we pass down the `metacardId` all the way down here even when it's `null`. Could we pass a `BiFunction<InputTransformer, InputStream, Metacard>` instead and just call it here?

In other words, the `transform(HarvestedResource, String)` method signature could look something like:
```
private Metacard transform(HarvestedResource harvestedResource,
    BiFunction<InputTransformer, InputStream, Metacard> transformFunction)
```
we could have the `public Metacard transform...()` methods call it like this:
```
public Metacard transformHarvestedResource(HarvestedResource harvestedResource) {
  return transform(harvestedResource, (it, is) -> it.transform(is));
}
```
or
```
public Metacard transformHarvestedResource(HarvestedResource resource, String metacardId) {
  return transform(harvestedResource, (it, is) -> it.transform(is, metacardId));
}
```
and have this method simply do:
```
private Optional<Metacard> doTransform(
     InputTransformer inputTransformer,
     InputStream is,
     BiFunction<InputTransformer, InputStream, Metacard> transformFunction,
     String resourceName) {
  try {
    return Optional.of(transformFunction.apply(inputTransformer, is));
  }
```
?